pages/blog/json-schema-beyond-validation.md

```diff
@@ -0,0 +1,152 @@
+
+However, in tooling, many times you do not want to validate data, but to represent the data in a structured manner so it is easier to interact with, such as classes that represent a message payload.
+
+## Quick intro to JSON Schema
+
+Let's try and take a look at an example. Given the following, we have a schema representing the validation rules that our data should comply with. 
+
+<figure>
+  <img src="../img/posts/jonaslagoni/json-schema-beyond-validation/json-schema-process.webp" title="JSON Schema validation process" alt="Shows the overall JSON Schema validation process of how a JSON Schema and some data is validated against each other."/>
+  <figcaption className="text-center text-gray-400 text-sm">Displays the overall process of validating data using JSON Schema.</figcaption>
+</figure>
+
+The JSON Schema defines that the JSON data should be an object, which requires a property called `someRequiredProperty` to always be present and an optional property called `someOptionalProperty`. `someRequiredProperty` should validate against an integer and `someOptionalProperty` against an arbitrary string. The schema also dictates that no additional properties (`"additionalProperties": false`) may be allowed. There is also some metadata defined, called `$id` and `$schema`, but they are not important for this example.
+
+If we then take a look at the example [data instances](https://datatracker.ietf.org/doc/html/draft-handrews-json-schema-01#section-4.2), the first one contains the required property, and the second one both the required and the optional property.
+
+The data and the JSON Schema can then together, validate whether the data is an instance of the schema, i.e., validate if the data comply with the validation rules and give a simple true or false statement if they are compatible.
+
+JSON Schema is an extremely powerful tool which allows you to create complex validation rules for data and is the standard specification used in not only AsyncAPI but also OpenAPI.
+
+## The problems of using JSON Schema as data definitions
+
+Many of the JSON Schema keywords are for [JSON instance validation](https://datatracker.ietf.org/doc/html/draft-handrews-json-schema-validation-01), which means specifying validation rules that data should comply with. However, what if you wanted to know the definition of the data rather than what it should validate against? That is currently not something JSON Schema provides to you, even though it is such an important part of tooling. Let's deep dive a bit into JSON Schema and see where some of the complexity lies.
+
+However, interpreting data definition from a JSON Schema is not always complex. For our previous example, we can almost interpret it as is. If I wanted a class in TypeScript that represented the data, it could look something like this (gonna use TS syntax as examples throughout). Notice how we use the `$id` keyword to define the name of such a class.
```

Comment:
after reading `Let's deep dive a bit into JSON Schema and see where some of the complexity lies.`, seeing word `However` somehow got me confused. We talk about complexity, then want to deep dive and all suddenly we say it is not always complex.

maybe instead of `However` write something like `First let me show you how to interpret data definition from a trivial schema`

Reply:
```suggestion
However, interpreting data definition from a JSON Schema is not always complex. For our previous example, we can almost interpret it as is. If I wanted a class in TypeScript that represented the data, it could look something like this (gonna use TS syntax as examples throughout). Notice how I use the `$id` keyword to define the name of such a class.
```

just a suggestion that you would have to apply to entire article, I just prefer first person content, not plural. Up to you

Reply:
Is the following better?
```
... First, let me show you how to interpret data definition from the example above for then to move into a more complex JSON Schema to further highlight the complexity. 

Interpreting data definition ...
```

Reply:
```
First, let me show you how to interpret data definition from the above example for then to move into a more complex JSON Schema. 
```

`to further highlight the complexity` no needed as  `complex JSON Schema.` is clear in that context + the sentence would get complicated

Reply:
> just a suggestion that you would have to apply to entire article, I just prefer first person content, not plural. Up to you

For many of the contexts, I don't feel like the first-person perspective makes sense.